Process does not terminate when UI is closed

I just installed GRIP-0.3.0.exe on Windows 7. When I terminate the program, either by selecting File => Exit or by clicking on the red X in the upper-right corner of the window, the UI closes but the process does not terminate. I am forced to terminate the GRIP.exe process using Windows Task Manager.

The exception that the program is responding to is an IO Exception.

```java
java.lang.RuntimeException: java.io.IOException: Cannot run program "$HOME\AppData\Local\GRIP\runtime/bin/java": CreateProcess error=2, The system cannot find the file specified
	at org.bytedeco.javacpp.Loader$2.run(Loader.java:645)
Caused by: java.io.IOException: Cannot run program "$HOME\AppData\Local\GRIP\runtime/bin/java": CreateProcess error=2, The system cannot find the file specified
	at java.lang.ProcessBuilder.start(ProcessBuilder.java:1048)
	at org.bytedeco.javacpp.Loader$2.run(Loader.java:643)
Caused by: java.io.IOException: CreateProcess error=2, The system cannot find the file specified
	at java.lang.ProcessImpl.create(Native Method)
	at java.lang.ProcessImpl.<init>(ProcessImpl.java:386)
	at java.lang.ProcessImpl.start(ProcessImpl.java:137)
	at java.lang.ProcessBuilder.start(ProcessBuilder.java:1029)
	... 1 more
```

--- a/core/src/main/java/edu/wpi/grip/core/events/UnexpectedThrowableEvent.java
+++ b/core/src/main/java/edu/wpi/grip/core/events/UnexpectedThrowableEvent.java
@@ -1,5 +1,7 @@
 package edu.wpi.grip.core.events;
 
+import edu.wpi.grip.core.util.SafeShutdown;
+
 import java.util.logging.Level;
 import java.util.logging.Logger;
 
@@ -50,7 +52,7 @@ public final class UnexpectedThrowableEvent {
             try {
                 logger.log(Level.SEVERE, "Failed to handle safely", throwable);
             } finally {
-                System.exit(1);
+                SafeShutdown.exit(1);
             }
         } finally {
             shutdownIfFatal();
@@ -67,7 +69,7 @@ public final class UnexpectedThrowableEvent {
                 logger.log(Level.SEVERE, "Shutting down from error", throwable);
             } finally {
                 // If all else fails then shutdown
-                System.exit(1);
+                SafeShutdown.exit(1);
             }
         }
     }

--- a/core/src/main/java/edu/wpi/grip/core/util/SafeShutdown.java
+++ b/core/src/main/java/edu/wpi/grip/core/util/SafeShutdown.java
@@ -1,0 +1,84 @@
+package edu.wpi.grip.core.util;
+
+/**
+ * This class should be used to shutdown GRIP safely.
+ * This is because shutdown hooks may throw exceptions, as such
+ * we need to know if the application being is shutdown.
+ */
+public final class SafeShutdown {
+
+    public static volatile boolean stopping = false;
+
+    static {
+        /*
+         * Shutdown hook run order is non-deterministic but this increases our likelihood of flagging a shutdown
+         * that we can't control.
+         */
+        Runtime.getRuntime().addShutdownHook(new Thread() {
+            @Override
+            public void run() {
+                stopping = true;
+            }
+        });
+    }
+
+    /**
+     * Shutdown's the VM in such a way that flags that the vm is stopping.
+     * This is so that we don't run the normal exception handling code when
+     * shutting down the application.
+     *
+     * @param statusCode exit status.
+     * @param hook       The hook to run before the System shutdown.
+     *                   This will be run after stopping has been flagged true.
+     *                   This is nullable.
+     * @see System#exit(int)
+     */
+    public static final void exit(int statusCode, PreSystemExitHook hook) {
+        flagStopping();
+        try {
+            if (hook != null) {
+                hook.run();
+            }
+        } finally {
+            System.exit(statusCode);
+        }
+
+    }
+
+    /**
+     * Helper method that passes null as the PreSystemExitHook.
+     *
+     * @param statusCode exit status.
+     * @see #exit(int)
+     */
+    public static final void exit(int statusCode) {
+        exit(statusCode, null);
+    }
+
+
+    /**
+     * HACK!
+     * Shutdown hooks can throw exceptions.
+     * On Windows, the static method after {@link org.bytedeco.javacpp.Loader#loadLibrary} throws such an exception
+     * in a shutdown hook.
+     *
+     * @return True if if the application is shutting down.
+     * @see <a href="https://github.com/WPIRoboticsProjects/GRIP/issues/297">GRIP Issue</a>
+     * @see <a href="https://github.com/bytedeco/javacpp/issues/60">Bytedeco issue</a>
+     */
+    public static final boolean isStopping() {
+        return stopping;
+    }
+
+    /**
+     * Flags that the application is shutting down.
+     */
+    public static void flagStopping() {
+        stopping = true;
+    }
+
+    @FunctionalInterface
+    public interface PreSystemExitHook {
+        void run();
+    }
+}

--- a/ui/src/main/java/edu/wpi/grip/ui/Main.java
+++ b/ui/src/main/java/edu/wpi/grip/ui/Main.java
@@ -9,6 +9,7 @@ import edu.wpi.grip.core.GRIPCoreModule;
 import edu.wpi.grip.core.Palette;
 import edu.wpi.grip.core.events.UnexpectedThrowableEvent;
 import edu.wpi.grip.core.operations.Operations;
+import edu.wpi.grip.core.util.SafeShutdown;
 import edu.wpi.grip.generated.CVOperations;
 import edu.wpi.grip.ui.util.DPIUtility;
 import javafx.application.Application;
@@ -31,9 +32,6 @@ public class Main extends Application {
     private Palette palette;
     @Inject
     private Logger logger;
-
-    private volatile boolean stopping = false;
-
 
     protected final Injector injector = Guice.createInjector(new GRIPCoreModule(), new GRIPUIModule());
 
@@ -62,8 +60,9 @@ public class Main extends Application {
         CVOperations.addOperations(eventBus);
 
         stage.setOnCloseRequest((event) -> {
-            Platform.exit();
-            System.exit(0);
+            // If this isn't here this can cause a deadlock on windows
+            // See issue #297
+            SafeShutdown.exit(0, Platform::exit);
         });
 
         stage.setTitle("GRIP Computer Vision Engine");
@@ -74,30 +73,35 @@ public class Main extends Application {
     }
 
     public void stop() {
-        stopping = true;
+        SafeShutdown.flagStopping();
     }
 
     @Subscribe
     @SuppressWarnings("PMD.AvoidCatchingThrowable")
     public final void onUnexpectedThrowableEvent(UnexpectedThrowableEvent event) {
         event.handleSafely((throwable, message, isFatal) -> {
-            // This should still use PlatformImpl
-            if (!stopping) {
+            // Check this so we can avoid entering the the platform wait
+            // if the program is shutting down.
+            if (!SafeShutdown.isStopping()) {
+                // This should still use PlatformImpl
                 PlatformImpl.runAndWait(() -> {
                     // WARNING! Do not post any events from within this! It could result in a deadlock!
                     synchronized (this.dialogLock) {
-                        try {
-                            // Don't create more than one exception dialog at the same time
-                            final ExceptionAlert exceptionAlert = new ExceptionAlert(root, throwable, message, isFatal, getHostServices());
-                            exceptionAlert.setInitialFocus();
-                            exceptionAlert.showAndWait();
-                        } catch (Throwable e) {
-                            // Well in this case something has gone very, very wrong
-                            // We don't want to create a feedback loop either.
+                        // Check again because the value could have been changed while waiting for the javafx thread to run.
+                        if (!SafeShutdown.isStopping()) {
                             try {
-                                logger.log(Level.SEVERE, "Failed to show exception alert", e);
-                            } finally {
-                                System.exit(1); // Ensure we shut down the application if we get an exception
+                                // Don't create more than one exception dialog at the same time
+                                final ExceptionAlert exceptionAlert = new ExceptionAlert(root, throwable, message, isFatal, getHostServices());
+                                exceptionAlert.setInitialFocus();
+                                exceptionAlert.showAndWait();
+                            } catch (Throwable e) {
+                                // Well in this case something has gone very, very wrong
+                                // We don't want to create a feedback loop either.
+                                try {
+                                    logger.log(Level.SEVERE, "Failed to show exception alert", e);
+                                } finally {
+                                    SafeShutdown.exit(1); // Ensure we shut down the application if we get an exception
+                                }
                             }
                         }
                     }

